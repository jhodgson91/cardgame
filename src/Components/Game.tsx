import React, { Component } from 'react';
import Hero from '../Components/Hero'
import House from '../Components/House'
import Card from '../Components/Card'
import Player from '../Components/Player'
import PlayerWrapper from '../Components/PlayerWrapper'
import './Main.scss'
import Deck from '../Objects/Deck'
import * as api from "../api"

//Define props types
export interface Props {

}

//Define state types
interface State {
  isReady : boolean;
  deck : Deck;
  cardInit : number;
  house : string;
  p1 : string;
  p2 : string;
}

export default class Game extends React.Component<Props, State> {
  
  constructor(props: Props) {
    super(props);
    this.state = {
      isReady: false,
      deck: new Deck(),
      cardInit: 26,
      house: 'house',
      p1: 'p1',
      p2: 'p2'
    };
    this.playCard = this.playCard.bind(this);
    this.cardViews = this.cardViews.bind(this);
  }
  
  async playCard(player: string) {
    if (this.state.isReady) {
      let d = this.state.deck;
      let main = this.state.house;
      let card = await d.piles[player].drawCardFrom('top');
      await d.piles[main].add([card]);
      
      this.setState({
        deck: d
      })
    }
  } 
  
  cardViews(pile: string) {
    let view = new Array<any>();
    if (this.state.isReady) {
      this.state.deck.piles[pile].cards.forEach(card => {
        view.push(
          <Card key={card.code} src={card.image}/>
        )
      })
    }
    return view;
  }
  
  async componentDidMount() {
    let d = await api.getDeck()
    if (d) {
      Promise.all([
        await d.newPile(this.state.house),
        await d.drawIntoPile(this.state.p1, this.state.cardInit),
        await d.drawIntoPile(this.state.p2, this.state.cardInit),
      ])
      this.setState({
        isReady: true,
        deck: d
      })
    }
  }

  render(): React.ReactNode {
    
    const house = this.state.house;
    const p1 = this.state.p1;
    const p2 = this.state.p2;
    
    return (
      <div className="cell">
        <Hero title={this.state.isReady && this.state.deck.id}/>
        <House title="House deck" cards={this.cardViews(house)}/>                               
        <PlayerWrapper 
          title="These are players"
          players={
            <div className="grid-x grid-margin-x align-center text-center"> 
              <Player title={p1} playCard={ () => {this.playCard(p1)} } cards={this.cardViews(p1)}/>
              <Player title={p2} playCard={ () => {this.playCard(p2)} } cards={this.cardViews(p2)}/>
            </div>
          }
        />
      </div>
    )
  }
}
 