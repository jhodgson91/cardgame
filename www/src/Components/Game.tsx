import * as React from 'react';
import './Main.scss';
import Deck from '../Objects/Deck';
import * as api from '../api';
import Hero from './Hero';
import PlayerWrapper from './PlayerWrapper';
import Player from './Player';
import Card from './Card';

//Define props types
export interface Props {

}

interface playerName {
    name: string,
    readOnly: boolean,
    theme: string
}

//Define state types
interface State {
    isReady: boolean;
    deck: Deck;
    drawFrom: 'top' | 'bottom';
    cardInit: number;
    players: playerName[];
}

export default class Game extends React.Component<Props, State> {

    constructor(props: Props) {
        super(props);
        this.state = {
            isReady: false,
            deck: new Deck(),
            drawFrom: 'top',
            cardInit: 26,
            players: [
                { name: 'house', readOnly: true, theme: "house" },
                { name: 'p1', readOnly: false, theme: "p1" },
                { name: 'p2', readOnly: false, theme: "p2" }
            ]
        };
        this.playCard = this.playCard.bind(this);
        this.showHand = this.showHand.bind(this);
        this.getCards = this.getCards.bind(this);
    }

    async playCard(player: string) {
        if (this.state.isReady) {
            let d = this.state.deck;
            let p = this.state.players;
            let loc = this.state.drawFrom;
            let card = await d.piles[player].drawCardFrom(loc);
            //Move a card to the central deck
            await d.piles[p[0].name].add([card]);
            this.setState({
                deck: d
            })
            //Check snap rules
            let snapDeck = d.piles[p[0].name].cards;
            let last = snapDeck.length - 1;
            if (typeof snapDeck[last - 1] !== 'undefined') {
                if (snapDeck[last].value === snapDeck[last - 1].value) {
                    window.alert("SNAP");
                }
            }
        }
    }

    getCards(player: string) {
        let d = this.state.deck;
        if (this.state.isReady) {
            let cards = d.piles[player].cards
                .map(c =>
                    <Card
                        key={c.code.toString()}
                        image={c.image}
                        value={c.value}
                        suit={c.suit}
                        code={c.code}
                    />
                );
            return cards;
        }
    }

    showHand(typeOfPlayer: boolean) {
        let p = this.state.players;
        let d = this.state.deck;
        if (this.state.isReady) {
            let hand = p
                .filter(i => i.readOnly === typeOfPlayer)
                .map(i =>
                    <Player
                        title={i.name}
                        key={i.name.toString()}
                        readOnly={i.readOnly}
                        theme={i.theme}
                        playCard={() => { this.playCard(i.name) }}
                    >
                        {this.getCards(i.name)}
                    </Player>
                );
            return hand;
        }
    }

    async componentDidMount() {
        let d = await api.getDeck();
        let p = this.state.players;
        let num = this.state.cardInit;
        if (d) {
            //Call services to initiate game
            Promise.all([
                await d.newPile(p[0].name),
                await d.drawIntoPile(p[1].name, num),
                await d.drawIntoPile(p[2].name, num)
            ]);
            //Update state so game starts
            this.setState({
                isReady: true,
                deck: d
            });
        }
    }

    render(): React.ReactNode {

        const isReady: boolean = this.state.isReady;
        const ID: string = this.state.deck.id;
        const loadingText: string = "Loading...";
        const smallWidth: number = 6;
        const largeWidth: number = 12;
        const RO: boolean = true;
        const W: boolean = false;

        return (
            <div className="cell">
                {isReady ? <Hero title={ID} /> : <Hero title={loadingText} />}
                <PlayerWrapper title="This is the house" grid={largeWidth}>
                    {this.showHand(RO)}
                </PlayerWrapper>
                <PlayerWrapper title="These are players" grid={largeWidth}>
                    {this.showHand(W)}
                </PlayerWrapper>
            </div>
        )
    }
}
