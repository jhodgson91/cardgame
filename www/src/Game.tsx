import * as React from 'react';

import axios, { AxiosResponse } from 'axios';
import {
    CardValue, MoveRequest, GameResponse,
    CardCode, CardSuit, CardSelection,
    BASE_URL, NEW_GAME_ID, INVALID_GAME_ID
} from './api';


export class Card {
    code: CardCode;

    constructor(code: CardCode) {
        this.code = code;
    }

    get value(): CardValue {
        return (this.code as string)[0] as CardValue;
    }
    get suit(): CardSuit {
        return (this.code as string)[1] as CardSuit;
    }

    jsx(): React.ReactNode {
        return (
            <div>{this.code}</div>
        )
    }

    toString() {
        return this.code as string;
    }
}

export type Piles = { [name: string]: Card[] }

export interface GameProps {
    id: string,
    piles: { [name: string]: CardSelection }
}

export interface GameState {
    id: string,
    piles: Piles
}

export default class Game extends React.Component<GameProps, GameState> {
    static defaultProps: GameProps = {
        id: NEW_GAME_ID,
        piles: {}
    }

    state: GameState = {
        id: INVALID_GAME_ID,
        piles: {}
    }

    // Props defaults to NEW_GAME_ID - in which case we POST
    // Otherwise, do a GET on the requested ID
    async componentDidMount() {
        let op = this.props.id != NEW_GAME_ID
            ? axios.get(`${BASE_URL}/${this.props.id}`)
            : axios.post(`${BASE_URL}/${NEW_GAME_ID}`, this.props.piles);

        this.refresh(await op);
    }

    // Draw cards from->to using selection
    async draw(from: string, to: string, selection: CardSelection) {
        let body: MoveRequest = {
            source: from,
            selection: selection,
        }

        this.refresh(await axios.put<GameResponse>(this.url + `/${to}/`, body));
    }

    // Updates state with data from a GameResponse
    refresh(response: AxiosResponse<GameResponse>) {
        let { id, piles } = response.data;

        let newPiles: Piles = {};
        Object.keys(piles).forEach(key => {
            newPiles[key] = piles[key].map((code: CardCode) => new Card(code));
        });

        this.setState({
            id: id,
            piles: newPiles
        });
    }

    render(): React.ReactNode {
        return (
            <div>
                Hello! My id is {this.id}
                <div>I have these piles:</div>
                <ul>
                    {
                        Object.keys(this.piles).map((key) => {
                            return <li key={key}>{key}
                                <ul>
                                    {
                                        this.piles[key].map((card: Card) => {
                                            return <li key={card.code}>{card.jsx()}</li>
                                        })
                                    }
                                </ul>
                            </li>
                        })
                    }
                </ul>
            </div>
        )
    }

    get id(): string { return this.state.id; }
    get isReady(): boolean { return this.state.id != INVALID_GAME_ID }
    get piles(): Piles { return this.state.piles; }
    get url(): string { return BASE_URL + `/${this.state.id}`; }
}