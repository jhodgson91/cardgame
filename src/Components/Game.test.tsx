import * as React from 'react';
import * as enzyme from 'enzyme';
import Game from './Game';

it('renders all children and game is initialised without the API', () => {
  const game = enzyme.shallow(<Game />);
  expect(game.find('Hero')).toHaveLength(1);
  expect(game.find('House')).toHaveLength(1);
  expect(game.find('PlayerWrapper')).toHaveLength(1);
  expect(game.state().cardInit).toEqual(26);
  expect(game.state().house).toEqual('house');
  expect(game.state().p1).toEqual('p1');
  expect(game.state().p2).toEqual('p2');
  expect(game.state().isReady).toEqual(false);
});

it('initialises properly when component is mounted', () => {
  const game = enzyme.mount(<Game />);
  expect(game.find('Hero')).toHaveLength(1);
  expect(game.find('House')).toHaveLength(1);
  expect(game.find('PlayerWrapper')).toHaveLength(1);
  expect(game.state().cardInit).toEqual(26);
  expect(game.state().house).toEqual('house');
  expect(game.state().p1).toEqual('p1');
  expect(game.state().p2).toEqual('p2');
  expect(game.state().isReady).toEqual(false);
});