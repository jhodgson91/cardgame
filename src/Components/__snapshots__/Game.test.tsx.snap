// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the same as last time 1`] = `
<div
  className="cell"
>
  <section
    id="hero"
  >
    <div
      className="grid-container"
    >
      <div
        className="grid-x grid-padding-x text-center"
      >
        <div
          className="small-12"
        >
          <h1>
            Card game - 
            Loading...
          </h1>
        </div>
      </div>
    </div>
  </section>
  <section
    className="players"
  >
    <div
      className="grid-container"
    >
      <div
        className="grid-x grid-padding-x align-center"
      >
        <div
          className="cell small-12"
        >
          <h2>
            This is the house
          </h2>
          <h3>
            Loading...
          </h3>
        </div>
      </div>
    </div>
  </section>
  <section
    className="players"
  >
    <div
      className="grid-container"
    >
      <div
        className="grid-x grid-padding-x align-center"
      >
        <div
          className="cell small-12"
        >
          <h2>
            These are players
          </h2>
          <h3>
            Loading...
          </h3>
        </div>
      </div>
    </div>
  </section>
</div>
`;
